pkg/go/graph/weighted_graph_builder.go

```diff
@@ -0,0 +1,168 @@
+package graph
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"gonum.org/v1/gonum/graph"
+)
+
+type WeightedAuthorizationModelGraphBuilder struct {
+	graph.DirectedMultigraphBuilder
+}
+
+func (wb *WeightedAuthorizationModelGraphBuilder) AddEdgeAndUpdateNodes(edge *AuthorizationModelEdge) error {
+	fromNode := wb.Node(edge.From().ID())
+	toNode := wb.Node(edge.To().ID())
+	isNested := fromNode == toNode
+
+	// update "isNested" field for the node
+	if isNested {
+		selfNode, ok := fromNode.(*WeightedAuthorizationModelNode)
+		if !ok {
+			return fmt.Errorf("%w: could not cast to WeightedAuthorizationModelNode", ErrBuildingGraph)
+		}
+		selfNode.isNested = true
+	}
+
+	// add line
+	edge.Line = wb.NewLine(fromNode, toNode)
+	newWeightedEdge := &WeightedAuthorizationModelEdge{edge, make(WeightMap), isNested}
+	wb.DirectedMultigraphBuilder.SetLine(newWeightedEdge)
+
+	return nil
+}
+
+func (wb *WeightedAuthorizationModelGraphBuilder) AssignWeights() error {
+	seen := make(map[int64]struct{})
+	iterNodes := wb.Nodes()
+	for iterNodes.Next() {
+		nextNode, ok := iterNodes.Node().(*WeightedAuthorizationModelNode)
+		if !ok {
+			return fmt.Errorf("%w: could not cast to WeightedAuthorizationModelNode", ErrBuildingGraph)
+		}
+
+		err := wb.dfsToAssignWeights(nextNode, seen)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+//nolint:gocognit,cyclop
+func (wb *WeightedAuthorizationModelGraphBuilder) dfsToAssignWeights(curNode *WeightedAuthorizationModelNode, seen map[int64]struct{}) error {
```

Comment:
The complexity of this method is rather high.  Would it make sense to break this into smaller functions (perhaps one per major step - i.e., dfsAssignWeightsToNeighorEdges, dfsAssignWeightsToNested, etc.).  This also allows unit testing at a per method level.